### PR TITLE
[CARBONDATA-3687] Support writing non-transactional carbondata files through hive

### PR DESCRIPTION
**Pull request by kunal642**
### Why is this PR needed?
1. Support creating non-transactional carbon tables through hive.
2. Support writting data(insert into) in carbondata format through hive.
 
 ### What changes were proposed in this PR?
1. Implemented MapredCarbonOutputFormat for data writing
2. Supported Map data type.
    
 ### Does this PR introduce any user interface change?
 - No
 - Yes. (please explain the change and update document)

 ### Is any new testcase added?
 - No
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1669/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1676/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1682/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1689/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1693/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1697/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1699/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1704/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1712/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1747/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1793/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1790/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/7/

**Comment by ydvpankaj99** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/125/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1829/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1881/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/178/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/180/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1883/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/181/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1884/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/182/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1885/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/190/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1892/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/199/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1901/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/201/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1903/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/202/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1904/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/204/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1906/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/205/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/206/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1908/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/208/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1910/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/209/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1911/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/212/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1914/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/222/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1924/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/253/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/254/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1957/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/379/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2081/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/381/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2083/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/382/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/383/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/385/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2086/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/387/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2088/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/393/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2094/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/405/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2106/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/522/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2221/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/524/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2224/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/532/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2232/

**Comment by kunal642** [comment]
@jackylk @ravipesala @ajantha-bhat @akashrn5 Please review

**Comment by ajantha-bhat** [comment]
please rebase and update the detail description

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/AvroCarbonWriterTest.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
 import org.apache.carbondata.core.constants.CarbonCommonConstants;
 import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier;
 import org.apache.carbondata.core.metadata.datatype.DataTypes;
+import org.apache.carbondata.core.metadata.datatype.Field;
```
Comment (jackylk) [inline sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/AvroCarbonWriterTest.java]: Seems not required to add?

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CSVCarbonWriterTest.java [hunk 2]
```diff
@@ -34,6 +34,7 @@
 import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier;
 import org.apache.carbondata.core.metadata.datatype.DataTypes;
 import org.apache.carbondata.core.metadata.datatype.DecimalType;
+import org.apache.carbondata.core.metadata.datatype.Field;
```
Comment (jackylk) [inline sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CSVCarbonWriterTest.java]: Seems not required to add?

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java [hunk 3]
```diff
@@ -33,6 +33,7 @@
 import org.apache.carbondata.core.datastore.impl.FileFactory;
 import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier;
 import org.apache.carbondata.core.metadata.datatype.DataTypes;
+import org.apache.carbondata.core.metadata.datatype.Field;
```
Comment (jackylk) [inline sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonReaderTest.java]: Seems not required to add?

File: sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonSchemaReaderTest.java [hunk 4]
```diff
@@ -27,6 +27,8 @@
 import org.apache.carbondata.core.datastore.filesystem.CarbonFileFilter;
 import org.apache.carbondata.core.datastore.impl.FileFactory;
 import org.apache.carbondata.core.metadata.datatype.DataTypes;
+import org.apache.carbondata.core.metadata.datatype.Field;
```
Comment (jackylk) [inline sdk/sdk/src/test/java/org/apache/carbondata/sdk/file/CarbonSchemaReaderTest.java]: Seems not required to add?

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/Field.java [hunk 5]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.carbondata.sdk.file;
+package org.apache.carbondata.core.metadata.datatype;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/Field.java]: Why move this package? Originally, this class is intended stay in SDK module so that user only uses API in carbondata-sdk module instead of calling functions in carbondata-core module

Comment (kunal642) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/Field.java]: Actually Field class already has the logic to parse complex child types, therefore instead of duplicating the code its better to move this class to core and use in both flows.

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2272/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/568/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/569/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2273/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/585/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2291/

File: processing/src/main/java/org/apache/carbondata/processing/loading/parser/GenericParser.java [hunk 6]
```diff
@@ -32,4 +32,6 @@
    */
   E parse(Object data);
 
+  E parseRaw(Object data);
```
Comment (jackylk) [inline processing/src/main/java/org/apache/carbondata/processing/loading/parser/GenericParser.java]: Add comment for this

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala [hunk 7]
```diff
@@ -758,7 +758,9 @@ class AddSegmentTestCase extends QueryTest with BeforeAndAfterAll {
     val writer = CarbonWriter.builder
       .outputPath(externalSegmentPath)
       .writtenBy("AddSegmentTestCase")
-      .withCsvInput(new Schema(fields))
+      .withSchemaFile(CarbonTablePath.getSchemaFilePath(CarbonEnv
```
Comment (jackylk) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala]: format it properly

File: integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      if (rs.getString(1).toLowerCase().contains(field.toLowerCase())) {
+        return rs.getString(2);
+      }
+    }
+    return "";
+  }
+
+  public boolean checkAnswer(ResultSet actual, ResultSet expected) throws SQLException {
+    Assert.assertEquals("Row Count Mismatch: ", expected.getFetchSize(), actual.getFetchSize());
+    while(expected.next()) {
```
Comment (jackylk) [inline integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java]: add space before (

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/HiveCarbonUtil.java [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+    CarbonLoadModel carbonLoadModel =
+        getCarbonLoadModel(tableName, databaseName, tablePath, sortColumns, columns.split(","),
+            columnTypes, configuration);
+    for (String delim : complexDelim.split(",")) {
+      carbonLoadModel.setComplexDelimiter(delim);
+    }
+    return carbonLoadModel;
+  }
+
+  public static CarbonLoadModel getCarbonLoadModel(String tableName, String databaseName,
```
Comment (jackylk) [inline integration/hive/src/main/java/org/apache/carbondata/hive/util/HiveCarbonUtil.java]: can you use CarbonLoadModelBuilder to build it

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/606/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2313/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/614/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2321/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/625/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2331/

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/Field.java (hunk 5 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/Field.java]: SDK guide need to update, many examples in guide have import statement of Field

File: integration/hive/pom.xml [hunk 10]
```diff
@@ -177,7 +188,7 @@
                 <!-- Note config is repeated in scalatest config -->
                 <configuration>
                     <reportsDirectory>${project.build.directory}/surefire-reports</reportsDirectory>
-                    <argLine>-Xmx3g -XX:MaxPermSize=512m -XX:ReservedCodeCacheSize=512m</argLine>
+                    <argLine>-Xmx3g -XX:MaxPermSize=512m -XX:ReservedCodeCacheSize=512m -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/tmp/kunal</argLine>
```
Comment (ajantha-bhat) [inline integration/hive/pom.xml]: please revert this

File: integration/hive/src/main/java/org/apache/carbondata/hive/CarbonHiveRecordReader.java [hunk 11]
```diff
@@ -106,6 +86,7 @@ private void initialize(InputSplit inputSplit, Configuration conf) throws IOExce
       int columnId = 0;
       for (int j = 0; j < arraySelectedColId.length; j++) {
         columnId = Integer.parseInt(arraySelectedColId[j]);
+
```
Comment (ajantha-bhat) [inline integration/hive/src/main/java/org/apache/carbondata/hive/CarbonHiveRecordReader.java]: revert this and remove local variable  `columnId`

File: integration/hive/pom.xml (hunk 10 above)
Comment (kunal642) [reply integration/hive/pom.xml]: done

File: integration/hive/src/main/java/org/apache/carbondata/hive/CarbonHiveRecordReader.java (hunk 11 above)
Comment (kunal642) [reply integration/hive/src/main/java/org/apache/carbondata/hive/CarbonHiveRecordReader.java]: done

File: integration/hive/src/main/java/org/apache/carbondata/hive/util/HiveCarbonUtil.java (hunk 9 above)
Comment (kunal642) [reply integration/hive/src/main/java/org/apache/carbondata/hive/util/HiveCarbonUtil.java]: done

File: integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java (hunk 8 above)
Comment (kunal642) [reply integration/hive/src/test/java/org/apache/carbondata/hive/HiveTestUtils.java]: done

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala (hunk 7 above)
Comment (kunal642) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala]: done

File: processing/src/main/java/org/apache/carbondata/processing/loading/parser/GenericParser.java (hunk 6 above)
Comment (kunal642) [reply processing/src/main/java/org/apache/carbondata/processing/loading/parser/GenericParser.java]: done

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java [hunk 12]
```diff
@@ -88,12 +89,23 @@ private static void populateCarbonTable(Configuration configuration, String path
...
+      String schemaPath = CarbonTablePath.getSchemaFilePath(validInputPath);
+      if (FileFactory.getCarbonFile(schemaPath).exists()) {
+        // read the schema file to get the absoluteTableIdentifier having the correct table id
+        // persisted in the schema
+        carbonTable = SchemaReader.readCarbonTableFromStore(absoluteTableIdentifier);
+      } else {
+        // InferSchema from data file
+        carbonTable = CarbonTable.buildFromTableInfo(SchemaReader
+            .inferSchema(absoluteTableIdentifier, false));
+        carbonTable.setTransactionalTable(true);
```
Comment (ajantha-bhat) [inline integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java]: If shema doesn't exists, will it be non-transactional table ?

can you add comment why it is set as transactional ?

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java [hunk 13]
```diff
@@ -18,43 +18,115 @@
...
 public class MapredCarbonOutputFormat<T> extends CarbonTableOutputFormat
-    implements HiveOutputFormat<Void, T> {
+    implements HiveOutputFormat<Void, T>, OutputFormat<Void, T> {
 
   @Override
   public RecordWriter<Void, T> getRecordWriter(FileSystem fileSystem, JobConf jobConf, String s,
-      Progressable progressable) {
-    return null;
+      Progressable progressable) throws IOException {
+    throw new RuntimeException("Should never be used");
```
Comment (ajantha-bhat) [inline integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: better to throw unsupported or more meaningful exception

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java [hunk 14]
```diff
@@ -18,43 +18,115 @@
...
   }
 
   @Override
-  public void checkOutputSpecs(FileSystem fileSystem, JobConf jobConf)
-      throws IOException {
-    org.apache.hadoop.mapreduce.JobContext jobContext = Job.getInstance(jobConf);
-    super.checkOutputSpecs(jobContext);
+  public void checkOutputSpecs(FileSystem fileSystem, JobConf jobConf) throws IOException {
   }
```
Comment (ajantha-bhat) [inline integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: can add some comment, why no functionality required here

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java [hunk 15]
```diff
@@ -18,43 +18,115 @@
...
-    return null;
+      Progressable progress) throws IOException {
+    CarbonLoadModel carbonLoadModel = null;
+    String encodedString = jc.get(LOAD_MODEL);
+    if (encodedString != null) {
+      carbonLoadModel =
+          (CarbonLoadModel) ObjectSerializationUtil.convertStringToObject(encodedString);
+    }
+    if (carbonLoadModel == null) {
+      carbonLoadModel = HiveCarbonUtil.getCarbonLoadModel(tableProperties, jc);
```
Comment (ajantha-bhat) [inline integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: better to keep it in `else` case of `if (encodedString != null)`

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java [hunk 16]
```diff
@@ -18,43 +18,115 @@
...
+        super.getRecordWriter(context);
+    return new FileSinkOperator.RecordWriter() {
+      @Override
+      public void write(Writable writable) throws IOException {
+        try {
+          ObjectArrayWritable objectArrayWritable = new ObjectArrayWritable();
+          if (isHivePartitionedTable) {
+            Object[] actualRow = ((CarbonHiveRow) writable).getData();
+            Object[] newData = Arrays.copyOf(actualRow, actualRow.length + partitionColumn);
+            String[] partitionValues = finalOutputPath.substring(tablePath.length()).split("/");
```
Comment (ajantha-bhat) [inline integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: what if the partition value itself contains "/" or "=" ? can happen if partition table is string column !

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java (hunk 12 above)
Comment (kunal642) [reply integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonInputFormat.java]: it should be false.. changed it

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java (hunk 13 above)
Comment (kunal642) [reply integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: it do not matter because this API is not called

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java (hunk 14 above)
Comment (kunal642) [reply integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: not needed for us

File: integration/hive/src/test/java/org/apache/carbondata/hive/HiveCarbonTest.java [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+public class HiveCarbonTest extends HiveTestUtils {
+
+  private static Statement statement;
+
+  @BeforeClass
+  public static void setup() throws Exception {
+    CarbonProperties.getInstance().addProperty(CarbonCommonConstants.ENABLE_OFFHEAP_SORT_DEFAULT, "false");
+    CarbonProperties.getInstance().addProperty(CarbonCommonConstants.ENABLE_UNSAFE_SORT, "false");
+    CarbonProperties.getInstance().addProperty(CarbonCommonConstants.CARBON_WRITTEN_BY_APPNAME, "hive");
```
Comment (ajantha-bhat) [inline integration/hive/src/test/java/org/apache/carbondata/hive/HiveCarbonTest.java]: Please set back these values to default at the end of test case, to avoid impact on existing test case.

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java (hunk 16 above)
Comment (kunal642) [reply integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: / in partition would be converted to '%2' when creating the partition folder, so it wount fail.

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala [hunk 18]
```diff
@@ -758,7 +758,9 @@ class AddSegmentTestCase extends QueryTest with BeforeAndAfterAll {
     val writer = CarbonWriter.builder
       .outputPath(externalSegmentPath)
       .writtenBy("AddSegmentTestCase")
-      .withCsvInput(new Schema(fields))
+      .withSchemaFile(CarbonTablePath.getSchemaFilePath(CarbonEnv.getCarbonTable(None,
```
Comment (ajantha-bhat) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala]: why this change ? now above fields() become unused. revert it

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java (hunk 15 above)
Comment (kunal642) [reply integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/631/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2338/

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala (hunk 18 above)
Comment (kunal642) [reply integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/addsegment/AddSegmentTestCase.scala]: reverted

File: integration/hive/src/test/java/org/apache/carbondata/hive/HiveCarbonTest.java (hunk 17 above)
Comment (kunal642) [reply integration/hive/src/test/java/org/apache/carbondata/hive/HiveCarbonTest.java]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/633/

File: integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java (hunk 16 above)
Comment (ajantha-bhat) [reply integration/hive/src/main/java/org/apache/carbondata/hive/MapredCarbonOutputFormat.java]: yeah ok, I remember now

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2340/

**Comment by ajantha-bhat** [comment]
LGTM

**Comment by jackylk** [comment]
Better not to move the `Field` class in sdk to core, since this is SDK API that many users have used already. Otherwise they can not compile

**Comment by kunal642** [comment]
@ajantha-bhat @jackylk it is for a major version release, i think its ok to change some API for this version.
It will be better than duplicate code.

**Comment by jackylk** [comment]
ok, please put a note in this class to notify the user

**Comment by ajantha-bhat** [comment]
can we still keep same package name inside core? so user need not have to change code ?

**Comment by kunal642** [comment]
@ajantha-bhat it dosent make sense to make a package names 'sdk' in core module. I think it is okay like this, we can put this as a breaking change.

**Comment by jackylk** [comment]
ok, please remember to mention this in 2.0's release note

**Comment by jackylk** [comment]
LGTM
